torch/csrc/distributed/c10d/TraceUtils.h

```diff
@@ -406,6 +408,36 @@ struct NCCLTraceBuffer {
     return id_++;
   }
 
+  void retire(Entry& r) {
+    r.retired_ = true;
```

Comment:
can you document the assumptions for this function? I think its important that the work obj is still alive when called, which is guaranteed by the call-site inside the workCleanupLoop right before deletion can happen.  (but i think it's not guaranteed by the other call-site inside dump())

Reply:
I split it in `update_state` which does only what it is named. And then put the stuff that changes other fields directly into the callers.